Bugfix for erlang maps encoding/decoding when maps_unset_optional is set to omitted

wolfpack9377 opened the pull request:
When encoding/decoding as maps (and maps_unset_optional set to omitted), the merge message function clause should not treat repeated fields as
required.

tomas-abrahamsson commented:
So what this does is on encode and merge, it also treats omitted repeated fields as if they had been given as an empty list?

Could you say something about the purpose/use-case to motivate (somewhat) increased complexity?

Hmm.. The generated code fails to compile for:
```protobuf
message m1 { map<string,fixed32> a = 1; }
```
with the following output:
```
m1.erl:445: variable 'F1' is unbound
m1.erl:444: Warning: variable 'M' is unused
```
It seems to be the code for generating the verifier that needs updating. The [`gpb_compile_maps_tests:map_type_test()`](https://github.com/tomas-abrahamsson/gpb/blob/3.27.2/test/gpb_compile_maps_tests.erl#L160) currently succeeds, but will fail correspondingly if one adds `{maps_unset_optional,omitted}` to the option-list. (Test coverage here lacks a bit, I'll add it to my todo-list, but it would also be nice for new features to be covered.) I suppose the same applies to the [`nif_encode_decode_mapfields()`](https://github.com/tomas-abrahamsson/gpb/blob/3.27.2/test/gpb_compile_maps_tests.erl#L608) test as well.

wolfpack9377 commented:
I like maps because of the flexibility of the structure (especially compared to records) and I was frustrated by the requirement to always define certain fields that I didn't care about at the application layer. It also seemed inconsistent since other types of optional fields were allowed to be omitted completely.

With the {maps_unset_optional,omitted} option, the current code generates function clauses which effectively assume the repeated fields are always present in the map (but not other optional fields):

```
e_msg_m9(#{optional_list := F3, optional_list_2 := F4} =
	     M,
	 Bin, TrUserData) ->
```
Definition file:
```
syntax = "proto3";

message m9 {
    string name = 1;
    bool flag = 2;
    repeated string optional_list = 3;
    repeated string optional_list_2 = 4;
}
```

so when you encode a map without that field, you get a function clause error:

```
test:encode_msg(#{name => "hello", flag => false}, 'm9').
** exception error: no function clause matching test:e_msg_m9(#{flag => false,name => "hello"},<<>>,undefined) (src/test.erl, line 47)
```

I will take another look at the test you mentioned.

tomas-abrahamsson commented:
I agree completely about the merits of maps. What I was thinking about was repeated. If you look at it from a `syntax="proto2"` perspective, there's `required`, `repeated` and `optional`. So I've never thought about `repeated` as optional, and thus it was sort-of implicit that it is always a list. But I see now that if one looks at it afresh from chiefly a proto3 perspective, with no proto2-aspects, it may make sense to see all kind of fields as omittable, even repeated. Perhaps especially so for maps with unset optionals omitted, since there is no default `= [ ]` like for records.

Given this rationale, omitting repeated fields should perhaps be possible only for proto3, if one was to be strict about it? I think I'm ok with repeated fields being omittable on encoding also for proto2, even though it makes less sense there.

Another thought is what an empty repeated field should decode to: should the field be omitted, for symmetry with encoding, or should it decode to an empty list. I think an empty list is the most practical return value. I know this is already the case in this PR. This paragraph is mostly for reference in case the question would pop up in the future.

Without tests and documentation, there is risk for this feature (for any feature really) to accidentally get lost in some refactorization. Perhaps a suitable doc place could be in the README.md, just after the table in [the data type mapping section](https://github.com/tomas-abrahamsson/gpb/#mapping-of-protocol-buffer-datatypes-to-erlang), and the gpb_compile_maps_tests.erl file contains tests that are run on Erlang versions with support for maps.

tomas-abrahamsson commented:
I just pushed a commit that addresses the compilation error and a test for it

wolfpack9377 commented:
Thanks - sorry for being missing in action - I was on vacation. I added a test for this issue in the gpb_compile_maps_test module and updated the README in the suggested section with the encode/decode behavior difference.

tomas-abrahamsson commented:
No problems at all.

Given
```protobuf
message m1 {
   repeated string f3 = 3;
}
```
and the options `maps,{maps_unset_optional,omitted}`, what should this return?
```erlang
1> Mod:merge(#{}, #{}, m1).
  -> ??
```
currently on the branch, it returns `#{}`, but I expected it to return `#{f3 => []}` analogous to how `Mod:decode_msg(<<>>, m1)` returns `#{f3 => []}`. I'm unsure if this would break any existing code. What do you think?

wolfpack9377 commented:
To me, it makes sense as is. However, I just started using the library so don't have a lot of legacy code to add insight to whether or not this would break existing use cases.

tomas-abrahamsson commented:
I pushed another commit to add test coverage for merge and verify for repeated fields, also covering `map<_,_>` which under the hood is also a repeated field.

If you are ok with it, then please squash the commits and I think it is ready to go.

tomas-abrahamsson commented:
After some thinking, I think I'm ok with the way merging is changed in this commit, since the property below still holds:

given some message, `x`, with a repeated field, then for some encoded instances, binaries `B1` and `B2`, the following should (and does) hold:
```erlang
M1 = Mod:decode_msg(B1, x),
M2 = Mod:decode_msg(B2, x),

Mod:decode_msg(<<B1/binary,B2/binary>>, x) =:= Mod:merge_msgs(M1, M2, x)
```

tomas-abrahamsson commented:
Thanks! Merged and included in 3.27.3 (closing)
